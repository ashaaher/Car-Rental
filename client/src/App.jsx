import React, { Fragment } from 'react';
import {
  CssBaseline,
  withStyles,
} from '@material-ui/core';

import { BrowserRouter as Router } from 'react-router-dom';
import AppWithRouterAccess from './AppWithRouterAccess';

const styles = theme => ({
  main: {
    padding: theme.spacing(0),
    [theme.breakpoints.down('xs')]: {
      padding: theme.spacing(2),
    },
  },
});

const App = ({classes}) => { 
  return (
    <Fragment>
      <CssBaseline />
      <main className={classes.main}>
        <Router>
          <AppWithRouterAccess/>
        </Router>
      </main>
    </Fragment>
  );
}

export default withStyles(styles)(App);