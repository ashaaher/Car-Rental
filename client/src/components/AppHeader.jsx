import React from 'react';
import { Link } from 'react-router-dom';

import {
  AppBar,
  Tabs,
  Tab,
  Toolbar,
  Typography,
  withStyles,
} from '@material-ui/core';
import { withOktaAuth } from '@okta/okta-react';
import LoginButton from "./LoginButton";
import { withRouter } from 'react-router-dom';

const styles = {
  flex: {
    flex: 1,
  },
  tabLink: {
    display: "flex",
    alignItems: "center",
  }
};

class AppHeader extends React.Component {
  state = {
    authenticated: false,
    user: null,
    role: "admin",
    value: 0,
  };

  componentDidUpdate() {
    this.checkAuthentication();
  }

  componentDidMount() {
    this.checkAuthentication();
  }

  async checkAuthentication() {
    const authenticated = this.props.authState ? this.props.authState.isAuthenticated : false;
    const user = this.props.authService ? await this.props.authService.getUser() : null;
    const role = user && user.groups && user.groups.includes("CarRentalApp Admin") ? "admin": "user";
    if (authenticated !== this.state.authenticated) {
      this.setState({
        authenticated,
        role
      })
    }
  }
  
  handleChange = (event, newValue) => {
    // event.preventDefault();
    this.setState({value: newValue});
  };
  render() {
    return (
      <AppBar position="static">
        <Toolbar>
          <Typography variant="h6" color="inherit">
            Car Rental System
          </Typography>
          {
            this.state.authenticated ?
              <>
                {
                  this.state.role === "admin" ?
                    <Tabs>
                      <Tab key="profile" label="Home" component={Link} className={this.props.classes.tabLink} to="/"></Tab>
                      <Tab key="vehicletype" label="Vehicle Types" component={Link} className={this.props.classes.tabLink} to="/vehicletype"></Tab>
                      <Tab key="settings" label="Settings" component={Link} className={this.props.classes.tabLink} to="/settings"></Tab>
                      <Tab key="locations" label="Locations" component={Link} className={this.props.classes.tabLink} to="/locations"></Tab>
                      <Tab key="vehicles" label="Vehicles" component={Link} className={this.props.classes.tabLink} to="/vehicles"></Tab>
                      <Tab key="users" label="Users" component={Link} className={this.props.classes.tabLink} to="/users"></Tab>

                    </Tabs> :
                    <Tabs>
                      <Tab key="home" label="Home" component={Link} className={this.props.classes.tabLink} to="/"></Tab>
                      <Tab key="customervehicles" label="Search Vehicles" component={Link} className={this.props.classes.tabLink} to="/customervehicles"></Tab>
                      <Tab key="bookings" label="My Bookings" component={Link} className={this.props.classes.tabLink} to="/bookings"></Tab>
                      <Tab key="profile" label="Profile" component={Link} className={this.props.classes.tabLink} to="/profile"></Tab>
                    </Tabs>
                }
              </>
              : null
          }


          <div className={this.props.classes.flex} />
          <LoginButton />
        </Toolbar>
      </AppBar>
    );
  }

}

export default withRouter(withOktaAuth(withStyles(styles)(AppHeader)));
