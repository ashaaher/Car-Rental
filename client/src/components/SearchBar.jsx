import React from 'react';
import {
    withStyles,
    CardContent,
    Button,
    TextField,
    Select,
    MenuItem,
} from '@material-ui/core';
import { withOktaAuth } from '@okta/okta-react';
import { compose } from 'recompose';
import { withRouter } from 'react-router-dom';
import { Form, Field } from 'react-final-form';
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import { required } from '../utils/validations';

const styles = theme => ({
    modal: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
    },
    modalCard: {
        width: '60%',
        maxWidth: 800
    },
    modalCardContent: {
        display: 'flex',
        flexDirection: 'row',
    },
    marginTop: {
        marginTop: theme.spacing(2),
    },
    modalField: {
        width: "25%",
        marginLeft: "20px"
    },
    dateInput: {
        marginLeft: "20px",
        textAlign: "center",
        height: "35px",
        marginTop: "15px",
        fontSize: "14px"
    },
    button: {
        marginLeft: '20px'
    }
});

class SearchBar extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        const { classes, onSearch, locations } = this.props;
        return <Form initialValues={{ locationId: 'all' }} onSubmit={onSearch}>
            {({ handleSubmit }) => (
                <form onSubmit={handleSubmit}>
                    <CardContent className={classes.modalCardContent}>
                        <Field name="searchText" validate={required}>
                            {({ input, meta }) => <>
                                <TextField className={classes.modalField} label="Name of the vehicle" {...input} />
                                {meta.error && meta.touched && <span style={{ color: "red" }}>*</span>}
                            </>}
                        </Field>
                        <Field name="locationId">
                            {({ input }) => <>
                                <Select
                                    className={classes.modalField}
                                    labelId="demo-simple-select-label"
                                    id="demo-simple-select"
                                    autoFocus
                                    {...input}
                                >
                                    {
                                        locations.map(loc => {
                                            return <MenuItem value={loc._id}>{loc.name}</MenuItem>
                                        })
                                    }
                                </Select>
                            </>}
                        </Field>
                        <Field name="startTime" validate={required}>
                            {({ input, meta }) => <>
                                <DatePicker
                                    className={classes.dateInput}
                                    selected={input.value}
                                    showTimeSelect
                                    timeFormat="HH:mm"
                                    timeIntervals={60}
                                    timeCaption="time"
                                    dateFormat="MMMM d, yyyy h:mm aa"
                                    placeholderText="Select pickup time"
                                    {...input}
                                />
                                {meta.error && meta.touched && <span style={{ color: "red" }}>*</span>}

                            </>}
                        </Field>
                        <Field name="endTime" validate={required}>
                            {({ input, meta }) => <>
                                <DatePicker
                                    className={classes.dateInput}
                                    selected={input.value}
                                    showTimeSelect
                                    timeFormat="HH:mm"
                                    timeIntervals={60}
                                    timeCaption="time"
                                    dateFormat="MMMM d, yyyy h:mm aa"
                                    placeholderText="Select return time"
                                    {...input}
                                />
                                {meta.error && meta.touched && <span style={{ color: "red" }}>*</span>}
                            </>}
                        </Field>
                        <Button variant="contained" className={classes.button} size="small" color="primary" type="submit">Search</Button>
                    </CardContent>
                </form>
            )}
        </Form>
    }
}

export default compose(
    withOktaAuth,
    withRouter,
    withStyles(styles),
)(SearchBar);
