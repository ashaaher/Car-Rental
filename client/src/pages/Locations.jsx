import React, { Fragment, Component } from 'react';
import { withOktaAuth } from '@okta/okta-react';
import { withRouter, Route, Redirect, Link } from 'react-router-dom';
import { compose } from 'recompose';
import {
  withStyles,
  Typography,
  Fab,
  Button,
  Paper,
  List,
  ListItem,
  ListItemText,
} from '@material-ui/core';
import { Add as AddIcon } from '@material-ui/icons';

import LocationEditor from '../components/editors/LocationEditor';
import ErrorSnackbar from '../components/ErrorSnackbar';
import apiClient, { patchBody } from '../libs/apiClient';
import Loading from '../components/Loading';

const styles = theme => ({
  posts: {
    marginTop: theme.spacing(2),
    padding: theme.spacing(3),
    maxHeight: '500px',
    overflow: 'auto',
    margin: '40px',
  },
  fab: {
    position: 'absolute',
    bottom: theme.spacing(3),
    right: theme.spacing(3),
    [theme.breakpoints.down('xs')]: {
      bottom: theme.spacing(2),
      right: theme.spacing(2),
    },
  },
});

class Locations extends Component {
  state = {
    loading: true,
    locations: [],
    error: null,
  };

  componentDidMount() {
    this.getLocations();
  }

  async getLocations() {
    try {
      const locations = await apiClient('/api/locations', { method: "GET" });
      this.setState({
        loading: false,
        locations
      });
    } catch (error) {
      this.setState({loading: false, error })
    }

  }

  saveLocation = async (data) => {
    try {
      let reqBody = {
        name: data.name,
        address: data.address,
        capacity: data.capacity
      }
      if(data._id) {
        reqBody = patchBody(reqBody);
      }
      const endpoint = data._id ? `/api/locations/${data._id}` : '/api/locations';
      await apiClient(
        endpoint,
        {
          method: data._id ? "PATCH" : "POST",
          body: JSON.stringify(reqBody)
        }
      );
    } catch (error) {
      this.setState({ error })
    }
    this.props.history.goBack();
    this.getLocations();
  }

  deleteLocation = async(id) => {
    if (window.confirm(`Are you sure you want to delete?`)) {
      try {
        await apiClient(
          `/api/locations/${id}`,
          {
            method: "DELETE",
          }
        )
        await this.getLocations();
      } catch (error) {
        this.setState({ error })
      }
      
    }
    
  } 

  renderEditor = ({ match: { params: { id } } }) => {
    if (this.state.loading) return null;
    let locationV = this.state.locations.find(v => v._id === id);

    if (locationV) {
      return <LocationEditor locationV={locationV} onSave={this.saveLocation} />;
    }

    if (!locationV && id !== 'new') return <Redirect to="/locations" />;

    return <LocationEditor locationV={locationV} onSave={this.saveLocation} />;
  };

  render() {
    if (this.props.authState.isPending || this.state.loading) {
      return <Loading />;
    }
    const { classes } = this.props;

    return (
      <Fragment>
        <Typography variant="h5">Locations</Typography>
        {this.state.locations.length > 0 ? (
          <Paper elevation={1} className={classes.posts}>
            <List>
              {this.state.locations.map(location => {
                return <ListItem key={location._id} button component={Link} to={`/locations/${location._id}`}>
                  <ListItemText
                    primary={location.name}
                    secondary={`Capacity : ${location.capacity}`}
                  />
                  <ListItemText
                    primary={`Address: ${location.address}`}
                  />
                  <Button size="small" color="primary" onClick={(event) => {
                    event.preventDefault();
                    this.deleteLocation(location._id)
                  }}>Delete</Button>
                </ListItem>
              })}
            </List>
          </Paper>
        ) : (
            !this.state.loading && <Typography variant="subtitle1">No locations to display</Typography>
          )}
        <Fab
          color="secondary"
          aria-label="add"
          className={classes.fab}
          component={Link}
          to="/locations/new"
        >
          <AddIcon />
        </Fab>
        <Route exact path="/locations/:id" render={this.renderEditor} />
        {this.state.error && (
          <ErrorSnackbar
            onClose={() => this.setState({ error: null })}
            message={this.state.error.message}
          />
        )}
      </Fragment>
    );
  }
};

export default compose(
  withOktaAuth,
  withRouter,
  withStyles(styles),
)(Locations);