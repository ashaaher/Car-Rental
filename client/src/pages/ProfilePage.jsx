import React from 'react';
import { withOktaAuth } from '@okta/okta-react';
import {
  Typography,
  Button,
  CardContent,
  CardActions
} from '@material-ui/core';
import apiClient from '../libs/apiClient';
import ErrorSnackbar from '../components/ErrorSnackbar';
import Loading from '../components/Loading';

export default withOktaAuth(
  class ProfilePage extends React.Component {
    constructor(props) {
      super(props);
      this.state = { loading: true, error: null, user: null };
      this.getCurrentUser = this.getCurrentUser.bind(this);
    }

    async getCurrentUser() {
      this.props.authService.getUser().then(user => this.setState({ user, loading: false }));
    }

    componentDidMount() {
      this.getCurrentUser();
    }

    async deleteUser() {
      try {
        await apiClient(`/api/users/${this.state.user.sub}`, { method: "DELETE" })
        this.props.authService.logout();
      } catch (error) {
        this.setState({ error })
      }
    }

    render() {
      if (this.props.authState.isPending || this.state.loading) {
        return <Loading />;
      }
      if (!this.state.user) return null;
      const startDate = new Date(this.state.user.updated_at * 1000);
      return (
        <section className="user-profile">
          <CardContent>
            <Typography color="textSecondary" gutterBottom>
              Name
            </Typography>
            <Typography variant="h5" gutterBottom>
              {this.state.user.name}
            </Typography>
            <Typography color="textSecondary" gutterBottom>
              Email
            </Typography>
            <Typography variant="h5" gutterBottom>
              {this.state.user.email}
            </Typography>
            <Typography color="textSecondary" gutterBottom>
              User Since
              </Typography>
            <Typography variant="h5" gutterBottom>
              {startDate.toLocaleString()}
            </Typography>
            <Typography color="textSecondary" gutterBottom>
              Membership Valid Upto
              </Typography>
            <Typography variant="h5" gutterBottom>
              {new Date(startDate.setMonth(startDate.getMonth() + 6)).toLocaleString()}
            </Typography>
            <CardActions>
              <Button variant="contained" onClick={() => { this.deleteUser() }}>Cancel Subscription</Button>
            </CardActions>
          </CardContent>
          {this.state.error && (
            <ErrorSnackbar
              onClose={() => this.setState({ error: null })}
              message={this.state.error.message}
            />
          )}
        </section>
      );
    }
  }
);
