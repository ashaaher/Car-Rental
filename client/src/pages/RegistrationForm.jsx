import React from 'react';
import OktaAuth from '@okta/okta-auth-js';
import { withOktaAuth } from '@okta/okta-react';

import config from '../app.config';
import {
  withStyles,
  CardContent,
  CardActions,
  Button,
  TextField,
  Card,
  Typography,
} from '@material-ui/core';
import { Form, Field } from 'react-final-form';
import ErrorSnackbar from '../components/ErrorSnackbar';
import { required, composeValidators, mustBeNumber, validateCardNumber } from '../utils/validations';

const styles = theme => ({
  modalCard: {
    width: '60%',
    maxWidth: 800,
    margin: 'auto'
  },
  modalCardContent: {
    display: 'flex',
    flexDirection: 'column',
  },
  marginTop: {
    marginTop: theme.spacing(2),
  },
});

export default withStyles(styles)(withOktaAuth(
  class RegistrationForm extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        sessionToken: null,
        error: null
      };

      this.oktaAuth = new OktaAuth({ issuer: config.issuer, pkce: false, cookies: config.cookies });
      this.checkAuthentication = this.checkAuthentication.bind(this);
      this.checkAuthentication();
      this.handleSubmit = this.handleSubmit.bind(this);
    }

    async checkAuthentication() {
      const sessionToken = await this.props.authService.getIdToken();
      if (sessionToken) {
        this.setState({ sessionToken });
      }
    }

    componentDidUpdate() {
      this.checkAuthentication();
    }

    async handleSubmit(data) {
      if (data.password !== data.repeatPassword) {
        this.setState({
          error: {
            message: "Passwords don't match"
          }
        })
        return;
      }
      fetch('/api/users', {
        method: 'POST',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(data)
      }).then(response => {
        if (!response.ok) {
          throw new Error('Registration failed');
        }
        this.oktaAuth
          .signIn({
            username: data.email,
            password: data.password
          })
          .then(res =>
            this.setState({
              sessionToken: res.sessionToken
            })
          );
      }).catch(err => {
        this.setState({ error: err })
      });
    }

    render() {
      const { classes } = this.props;
      if (this.state.sessionToken) {
        this.props.authService.redirect({ sessionToken: this.state.sessionToken });
        return null;
      }
      return <Form onSubmit={this.handleSubmit}>
        {({ handleSubmit }) => (<Card className={classes.modalCard}>
          <form onSubmit={handleSubmit}>
            <Typography variant="h5">Registration</Typography>
            <CardContent className={classes.modalCardContent}>
              <Field name="email" validate={required}>
                {({ input, meta }) => <>
                  <TextField label="Email"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="firstName" validate={required}>
                {({ input, meta }) => <>
                  <TextField label="First Name"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="lastName" validate={required}>
                {({ input, meta}) => <>
                  <TextField label="Last Name"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="password" validate={required}>
                {({ input,meta }) => <>
                  <TextField label="Password"  type="password" {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="repeatPassword" validate={required}>
                {({ input, meta }) => <>
                  <TextField label="Confirm Password" type="password" {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="address" validate={required}>
                {({ input, meta }) => <>
                <TextField label="Address" multiline
                  rows={4} {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                  </>}
              </Field>
              <Field name="licenceNumber" validate={required}>
                {({ input, meta }) => <>
                  <TextField label="Driving Licence Number"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="stateOfResidence" validate={required}>
                {({ input, meta }) => <>
                  <TextField label="Driving Licence State"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="nameOnCard" validate={required}>
                {({ input, meta }) => <>
                  <TextField label="Name on Card"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="cardNumber" validate={composeValidators(required, mustBeNumber, validateCardNumber)}>
                {({ input, meta }) => <>
                  <TextField label="Card Number"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="cvv" validate={composeValidators(required, mustBeNumber)}>
                {({ input, meta }) => <>
                  <TextField label="CVV"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="expirationDate" validate={required}>
                {({ input, meta }) => <>
                  <TextField label="Expiration Date (MM/YY)"  {...input} />
                  {meta.error && meta.touched && <span style={{color:"red"}}>{meta.error}</span>}
                </>}
              </Field>

            </CardContent>
            <CardActions>
              <Button variant="contained" color="primary" disableElevation type="submit">Register</Button>
            </CardActions>
            {this.state.error && (
              <ErrorSnackbar
                onClose={() => this.setState({ error: null })}
                message={this.state.error.message}
              />
            )}
          </form>
        </Card>)}
      </Form>

    }
  }
));
