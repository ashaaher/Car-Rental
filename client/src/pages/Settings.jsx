import React from 'react';
import {
  Typography,
} from '@material-ui/core';
import {
  withStyles,
  Card,
  CardContent,
  CardActions,
  Button,
  TextField,
  Paper
} from '@material-ui/core';
import { compose } from 'recompose';
import { withRouter } from 'react-router-dom';
import { Form, Field } from 'react-final-form';
import { withOktaAuth } from '@okta/okta-react';
import apiClient, { patchBody } from '../libs/apiClient';
import ErrorSnackbar from '../components/ErrorSnackbar';
import Loading from '../components/Loading';
import { required, composeValidators, mustBeNumber, minValue, maxValue } from '../utils/validations';


const styles = theme => ({
  card: {
    width: '90%',
    maxWidth: 500,
  },
  cardContent: {
    display: 'flex',
    flexDirection: 'column',
  },
  cardField: {
    marginTop: theme.spacing(2),
  },
  marginTop: {
    marginTop: theme.spacing(2),
  },
});


class Settings extends React.Component {

  state = {
    settings: null,
    error: null
  };

  componentDidMount() {
    this.getSettings();
  }

  async getSettings() {
    try {
      const [settings] = await apiClient('/api/settings', { method: "GET" })
      this.setState({
        settings
      })
    } catch (error) {
      this.setState({ error })
    }

  }

  handleSubmit = async (data) => {
    try {
      await apiClient('/api/settings', {
        method: "PUT",
        body: JSON.stringify(data)
      })
      await this.getSettings();
      alert("Settings saved successfully");
    } catch (error) {
      this.setState({ error })
    }
    
  }

  render() {
    if (this.props.authState.isPending || this.state.loading) {
      return <Loading />;
    }
    const { history, classes } = this.props;
    return <Typography variant="h4" align='center'>
      <Form initialValues={this.state.settings} onSubmit={this.handleSubmit}>
        {({ handleSubmit }) => (<Card className={classes.card}>
          <form onSubmit={handleSubmit}>
            <CardContent className={classes.cardContent}>
              <Paper className={classes.root}>
                <Typography variant="h5" component="h3">
                  Application Settings
                </Typography>
              </Paper>
              <Field name="membershipFee" validate={composeValidators(required, mustBeNumber, minValue(0))}>
                {({ input, meta }) => <>
                  <TextField className={classes.cardField} label="Membership Fee ($)" autoFocus {...input} />
                  {meta.error && meta.touched && <span style={{ color: "red" }}>{meta.error}</span>}
                </>}
              </Field>
              <Field name="lateFee" validate={composeValidators(required, mustBeNumber, minValue(0))}>
                {({ input, meta }) => <>
                  <TextField className={classes.cardField} label="Late Payment Fee ($)" autoFocus {...input} />
                  {meta.error && meta.touched && <span style={{ color: "red" }}>{meta.error}</span>}
                </>}
              </Field>
            </CardContent>
            <CardActions>
              <Button size="small" color="primary" type="submit">Save</Button>
              <Button size="small" onClick={() => history.goBack()}>Cancel</Button>
            </CardActions>
          </form>
        </Card>
        )}
      </Form>
      {this.state.error && (
        <ErrorSnackbar
          onClose={() => this.setState({ error: null })}
          message={this.state.error.message}
        />
      )}
    </Typography>
  }
}

export default compose(
  withOktaAuth,
  withRouter,
  withStyles(styles),
)(Settings);
