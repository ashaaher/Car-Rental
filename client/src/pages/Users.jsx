import React, { Component, Fragment } from 'react';
import { withOktaAuth } from '@okta/okta-react';
import { withRouter } from 'react-router-dom';
import {
  withStyles,
  Typography,
  Button,
  Paper,
  List,
  ListItem,
  ListItemText,
  ListItemSecondaryAction,
} from '@material-ui/core';
import { compose } from 'recompose';

import ErrorSnackbar from '../components/ErrorSnackbar';
import apiClient, { patchBody } from '../libs/apiClient';
import Loading from '../components/Loading';

const styles = theme => ({
  posts: {
    marginTop: theme.spacing(2),
    maxHeight: "580px",
    overflow: 'auto',
    margin: '50px'
  },
  fab: {
    position: 'absolute',
    bottom: theme.spacing(3),
    right: theme.spacing(3),
    [theme.breakpoints.down('xs')]: {
      bottom: theme.spacing(2),
      right: theme.spacing(2),
    },
  },
});

class Users extends Component {
  state = {
    loading: true,
    users: [],
    error: null,
  };

  componentDidMount() {
    this.getUsers();
  }

  async getUsers() {
    try {
      const users = await apiClient('/api/users', { method: "GET" });
      this.setState({
        loading: false,
        users
      });
    } catch (error) {
      this.setState({error})
    }
    
  }

  async revokeMembership(user) {
    if (window.confirm(`Are you sure you want to revoke user?`)) {
      await apiClient(
        `/api/users/${user.id}`,
        {
          method: "DELETE",
        }
      )
      await this.getUsers();
    }
  }

  render() {
    if (this.props.authState.isPending || this.state.loading) {
      return <Loading />;
    }
    const { classes } = this.props;

    return (
      <Fragment>
        <Typography variant="h5">Users</Typography>
        {this.state.users.length > 0 ? (
          <Paper elevation={1} className={classes.posts}>
            <List>
              {this.state.users.map(user => (
                <ListItem key={user.profile.email}>
                  <ListItemText
                    primary={user.profile.firstName}
                    secondary={user.profile.lastName}
                  />
                  <ListItemText
                    primary={user.profile.email}
                  />
                  <ListItemSecondaryAction>
                    <Button onClick={() => this.revokeMembership(user)} color="inherit">
                      Revoke
                    </Button>
                  </ListItemSecondaryAction>
                </ListItem>
              ))}
            </List>
          </Paper>
        ) : (
            !this.state.loading && <Typography variant="subtitle1">No users to display</Typography>
          )}
        {this.state.error && (
          <ErrorSnackbar
            onClose={() => this.setState({ error: null })}
            message={this.state.error.message}
          />
        )}
      </Fragment>
    );
  }
}

export default compose(
  withOktaAuth,
  withRouter,
  withStyles(styles),
)(Users);
