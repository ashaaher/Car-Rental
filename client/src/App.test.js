/**
 author: Vijaylaxmi Nagurkar
*/

import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from './App';
import Locations from './pages/Locations';
import ProfilePage from './pages/ProfilePage';
import RegistrationForm from './pages/RegistrationForm';
import Settings from './pages/Settings';
import Users from './pages/Users';
import Vehicles from './pages/Vehicles';
import VehicleType from './pages/VehicleType';
import AppHeader from './components/AppHeader';
import Loading from './components/Loading';
import LoginButton from './components/LoginButton';
import LoginForm from './components/LoginForm';
import Navigation from './components/Navigation';
import SearchBar from './components/SearchBar';






it('App renders without crashing enzyme', () => {
  shallow(<App />);
});

it('App matches snapshot', () => {
  const wrapper = shallow(<App />);
  expect(wrapper).toMatchSnapshot();
})
it('Locations Page matches snapshot', () => {
  const wrapper = shallow(<Locations />);
  expect(wrapper).toMatchSnapshot();
})

it('AppHeader Page matches snapshot', () => {
  const wrapper = shallow(<AppHeader />);
  expect(wrapper).toMatchSnapshot();
})
it('ProfilePage Page matches snapshot', () => {
  const wrapper = shallow(<ProfilePage />);
  expect(wrapper).toMatchSnapshot();
})
it('RegistrationForm Page matches snapshot', () => {
  const wrapper = shallow(<RegistrationForm />);
  expect(wrapper).toMatchSnapshot();
})
it('Settings Page matches snapshot', () => {
  const wrapper = shallow(<Settings />);
  expect(wrapper).toMatchSnapshot();
})
it('Users Page matches snapshot', () => {
  const wrapper = shallow(<Users />);
  expect(wrapper).toMatchSnapshot();
})
it('Vehicles Page matches snapshot', () => {
  const wrapper = shallow(<Vehicles />);
  expect(wrapper).toMatchSnapshot();
})
it('VehicleType Page matches snapshot', () => {
  const wrapper = shallow(<VehicleType />);
  expect(wrapper).toMatchSnapshot();
})
it('Loading Page matches snapshot', () => {
  const wrapper = shallow(<Loading />);
  expect(wrapper).toMatchSnapshot();
})
it('LoginButton Page matches snapshot', () => {
  const wrapper = shallow(<LoginButton />);
  expect(wrapper).toMatchSnapshot();
})
it('LoginForm Page matches snapshot', () => {
  const wrapper = shallow(<LoginForm />);
  expect(wrapper).toMatchSnapshot();
})
it('Navigation Page matches snapshot', () => {
  const wrapper = shallow(<Navigation />);
  expect(wrapper).toMatchSnapshot();
})
it('SearchBar Page matches snapshot', () => {
  const wrapper = shallow(<SearchBar />);
  expect(wrapper).toMatchSnapshot();
})
