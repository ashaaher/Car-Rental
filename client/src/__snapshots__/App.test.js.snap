// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `ShallowWrapper {}`;

exports[`AppHeader Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`Loading Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`Locations Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`LoginButton Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`LoginForm Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`Navigation Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`ProfilePage Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`RegistrationForm Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`SearchBar Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`Settings Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`Users Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`VehicleType Page matches snapshot 1`] = `ShallowWrapper {}`;

exports[`Vehicles Page matches snapshot 1`] = `ShallowWrapper {}`;
